transport/tlsfrag/buffer.go

```diff
@@ -0,0 +1,307 @@
+		m, e := w.Write(b.data[:b.len])
+		n = int64(m)
+		if err = e; err != nil {
+			return
+		}
+		// all bytes should have been written, by definition of Write method in io.Writer
+		if m != b.len {
+			err = io.ErrShortWrite
+		}
+	}
+	return
+}
+
+// HasFullyReceived returns whether a complete TLS Client Hello packet has been assembled.
+func (b *clientHelloBuffer) HasFullyReceived() bool {
+	return b.valid && b.recordLen > 0 && b.len >= b.recordLen+tlsRecordHeaderSize
+}
+
+// BytesOverlapped returns the number of bytes actually copied from the io.Reader in ReadFrom(r)
+// to io.Writer in WriteTo, ignoring any extra headers added by Split.
+//
+// Here's an example explaining it further:
+//
+//	_, _ := buf.ReadFrom([]byte{1,2})      // {1,2} are appended to buf
+//	rn, _ := buf.ReadFrom([]byte{3,4,5,6}) // rn == 3, {3,4,5} are appended to buf
+//	buf.Split(2)                           // will add some additional header bytes
+//	// now assume buf contains {1,2,h,h,h,h,h,3,4,5}
+//	wn, _ := buf.WriteTo(w)                // wn == 8, {1,2,h,h,h,h,h,3} are written to w
+//	n := buf.BytesOverlapped(rn, wn)       // n == 1, because only byte {3} comes from the last ReadFrom
+func (b *clientHelloBuffer) BytesOverlapped(rn, wn int64) int {
+	//   ndata = 12:  1 2 3 4 h h h h h 5 6 7
+	//       rn = 5:      | |           | | |
+	//       wn = 6:  | | | | | |
+	// overlap == 2:      ^ ^
+	//       wn & h:  x x x x | | N N N
+
+	if wn < int64(b.split) {
+		// add all 5 header bytes to wn when splitted and wn doesn't overlap with h
+		// if no splitting, this condition will never be satifsfied because wn always >= 0
+		wn += tlsRecordHeaderSize
+	} else if b.split > 0 && wn < int64(b.split+tlsRecordHeaderSize) {
+		// fill all non-overlapped h bytes to wn (bytes marked as N above) when wn partially overlaps with h
+		wn = int64(b.split + tlsRecordHeaderSize)
+	}
+
+	// now both wn and n contain either a 5-byte header or no header at all
+	// the header bytes get cancelled out in the subtraction (wn - ndata) below
+	// rn + wn = (left+overlap) + (overlap+right) = (left+overlap+right) + overlap = ndata + overlap
+	if overlap := int(rn) + int(wn) - b.len; overlap >= 0 {
+		return overlap
+	}
+	return 0
+}
+
+// Content returns the Client Hello packet content (without the 5 bytes header).
+// It might return an incomplete content, the caller needs to make sure HasFullyReceived before calling this function.
+func (b *clientHelloBuffer) Content() []byte {
+	if b.len <= tlsRecordHeaderSize {
+		return []byte{}
+	}
+	return b.data[tlsRecordHeaderSize:b.len]
+}
+
+// Split fragments the Client Hello packet into two TLS records at the specified 0-based splitBytes:
+// [:splitBytes] and [splitBytes:]. Any necessary headers will be added to this buffer.
+//
+// If the packet has already be splitted before, a non-nil error and returned.
+// If the split index is ≤ 0 or ≥ the total length, do nothing.
+func (b *clientHelloBuffer) Split(splitBytes int) error {
```

Comment:
As discussed offline, let's separate the record buffering and the split logic, so it's easier to understand and modify.